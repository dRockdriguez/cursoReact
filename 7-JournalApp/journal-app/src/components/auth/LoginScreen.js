import React from "react";
import { Link } from "react-router-dom";

export const LoginScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-2">Login</h3>

      <form>
        <input
          className="auth__input mb-1"
          type="email"
          placeholder="Email..."
          name="email"
          autoComplete="off"
        />
        
        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Password..."
          name="password"
        />

        <button className="btn btn-primary btn-block" type="submit">
          Go
        </button>

        <div className="auth__social-networks">
          <p className="mb-2">Login with Social Networks</p>

          <div className="google-btn">
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>

          <Link className="link mt-2" to="/auth/register">
            Create new account
          </Link>
        </div>
      </form>
    </>
  );
};
