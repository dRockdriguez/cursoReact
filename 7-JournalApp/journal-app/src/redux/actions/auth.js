import Swal from "sweetalert2";
import { types } from "../types/types";
import { googleAuthProvider } from "../../firebase/firebase-config";
import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import { finishLoading, startLoading } from "./ui";

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const { user } = await signInWithEmailAndPassword(auth, email, password);
      const { uid, displayName } = user;
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const {
        user: { uid, displayName },
      } = await signInWithPopup(auth, googleAuthProvider);
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startRegister = ({ email, password, name }) => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const { user } = await createUserWithEmailAndPassword(
        auth,
        email,
        password
      );
      await updateProfile(user, { displayName: name });
      const { uid, displayName } = user;
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      await signOut(auth);
      dispatch(logout());
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const logout = () => ({
  type: types.logout,
});

const showError = (message) => {
  Swal.fire({
    icon: "error",
    title: "Oops...",
    text: message,
  });
};
