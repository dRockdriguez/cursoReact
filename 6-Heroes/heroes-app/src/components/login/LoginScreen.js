import React, { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { types } from "../../types/types";
import { AuthContext } from "../../auth/AuthContext";

export const LoginScreen = () => {
  const navigate = useNavigate();
  const { dispatch } = useContext(AuthContext);

  const handleLogin = () => {
    const action = {
      type: types.login,
      payload: { name: "David" },
    };
    
    dispatch(action);

    navigate("/dashboard", {
      replace: true,
    });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>

      <hr />

      <button className="btn btn-primary" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
