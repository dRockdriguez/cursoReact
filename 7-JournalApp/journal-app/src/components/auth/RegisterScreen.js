import React from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { setError, removeError } from "../../redux/actions/ui";

export const RegisterScreen = () => {
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    confirm: "",
  });

  const { name, email, password, confirm } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    const valid = isValidForm();
    if (valid.valid) {
      dispatch(removeError());
    } else {
      dispatch(setError(valid.error));
    }
  };

  const isValidForm = () => {
    if (name.trim().length <= 0) {
      return {
        valid: false,
        error: "The name is invalid",
      };
    } else if (!validator.isEmail(email)) {
      return {
        valid: false,
        error: "The email is invalid",
      };
    } else if (password.length <= 5 || password !== confirm) {
      return {
        valid: false,
        error: "The password is invalid",
      };
    }
    return {
      valid: true,
    };
  };

  return (
    <>
      <h3 className="auth__title mb-2">Sign up</h3>

      <form onSubmit={handleRegister}>
        <div className="auth__alert-error">Hola</div>

        <input
          className="auth__input mb-1"
          type="name"
          placeholder="Name..."
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="email"
          placeholder="Email..."
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Password..."
          name="password"
          value={password}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Confirm password..."
          name="confirm"
          value={confirm}
          onChange={handleInputChange}
        />

        <button className="btn btn-primary btn-block mb-2" type="submit">
          Go
        </button>

        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  );
};
