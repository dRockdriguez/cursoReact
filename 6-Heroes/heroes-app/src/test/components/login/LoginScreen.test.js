import { LoginScreen } from "../../../components/login/LoginScreen";
import { mount } from "enzyme";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { AuthContext } from "../../../auth/AuthContext";
import { types } from "../../../types/types";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("LoginScreen", () => {
  const contextValue = {
    user: {
      logged: false,
    },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={["/login"]}>
        <Routes>
          <Route path="/login" element={<LoginScreen />}></Route>
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test("Debe mostrarse correctamente por defecto", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe lanzar el dispatch y navegaciÃ³n", () => {
    const handleClick = wrapper.find("button").prop("onClick");
    localStorage.setItem("lastPath", "/prueba");
    
    handleClick();

    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: { name: "David" },
    });

    expect(mockNavigate).toHaveBeenCalledWith("/prueba", { replace: true });

  });
});
