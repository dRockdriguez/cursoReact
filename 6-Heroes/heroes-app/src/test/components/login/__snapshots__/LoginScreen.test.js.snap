// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen Debe mostrarse correctamente por defecto 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
